HPCC-19216 Dynamically add <Properties> to ESPService and ESPBinding sections

kenrowland opened the pull request:
Add support to dynamically add <Properties> sections to
ESPService and ESPBinding sections in the environment

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

kenrowland commented:
@rpastrana @afishbeck Please review

This PR adds event processing for element creating that inserts XML (properties and authentication features) into dependent sections, namely ESPService and ESPBinding.

It also adds dependent attribute values. This handles default http and https port numbers for ESPBindings.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 66caf24c6066cfec9cd04f700124a3aba97013f8
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |85 |85 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |689 |689 |0 |
|test (roxie) |802 |802 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    6 sec (00:00:06) |  190 sec (00:03:10) |   56 sec (00:00:56) |    5 sec (00:00:05) |   21 sec (00:00:21) | 1109 sec (00:18:29) |   16 sec (00:00:16) | 1403 sec (00:23:23) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 2ca63e54e05bd382f8e4e51d806e6c67a8099167
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |85 |85 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |689 |689 |0 |
|test (roxie) |802 |802 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  184 sec (00:03:04) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1123 sec (00:18:43) |   16 sec (00:00:16) | 1438 sec (00:23:58) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 53b00ed86d1a768d4c3f2c04c00bc46dd7cca361
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |85 |85 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |689 |689 |0 |
|test (roxie) |802 |802 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  181 sec (00:03:01) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1144 sec (00:19:04) |   17 sec (00:00:17) | 1439 sec (00:23:59) |

afishbeck commented on configuration/config2/EnvironmentEvents.cpp:
```diff
@@ -0,0 +1,106 @@
+            {
+                if (!m_itemAttribute.empty())
+                {
+                    std::shared_ptr<EnvironmentValue> pItemAttr = pEventNode->getAttribute(m_itemAttribute);
```
Should cases like these be "unique_ptr" rather than shared_ptr?  If so, there are probably many cases throughout.

kenrowland replied:
In most cases the called method returns a shared_ptr since because the object already exists and most likely has a reference. Converting to a unique_ptr in these cases would not follow the architecture since unique ownership is not the intent. I will, however, keep an eye out for situations where a unique_ptr would suffice, such as where the method creates the object and persistence is not necessary.

afishbeck commented on configuration/config2/EnvironmentLoader.hpp:
```diff
@@ -0,0 +1,49 @@
+
+        EnvironmentLoader() { }
+        virtual ~EnvironmentLoader() { }
+        //void addPath(const std::shared_ptr<EnvironmentNode> pNode);
```
Not really a fan of keeping commented out code... are these functions going to be readded at some point?

kenrowland replied:
Methods were moved during some refactoring. I neglected to remove the commented lines. Lines now removed.

afishbeck commented on configuration/config2/SchemaItem.hpp:
```diff
@@ -73,6 +77,11 @@ class DECL_EXPORT SchemaItem : public std::enable_shared_from_this<SchemaItem>
         bool isHidden() const { return m_hidden; }
 
+        void setParent(const std::shared_ptr<SchemaItem> &parent) { m_pParent = parent; }
+        std::shared_ptr<const SchemaItem> findSchemaRoot() const;
```
I understand the implementation "walks" the parent nodes, but conceptually this doesn't seem like a "find" function,.. more like a "getSchemaRoot()" or similiar name.

kenrowland replied:
True, it is a get, find was carried over because it was written while I was implementing the find functionality. I changed it to getSchemaRoot. Similar change made to finding the root node in the environment.

afishbeck commented on configuration/config2/EnvironmentEvents.hpp:
```diff
@@ -0,0 +1,97 @@
+        void setItemAttributeName(const std::string &name) { m_itemAttribute = m_matchAttribute = name; }
+        void setMatchAttributeName(const std::string &name) { m_matchAttribute = name; }
+        virtual bool handleEvent(const std::string &envType, std::shared_ptr<EnvironmentNode> pEventNode);
+
```
Trivial, but lots of extra space.

afishbeck commented on configuration/config2/EnvironmentEvents.hpp:
```diff
@@ -0,0 +1,97 @@
+        virtual ~InsertEnvironmentDataCreateEvent() {}
+        void setEnvironmentInsertData(const std::string &envData) { m_envData = envData;  }
+        void setMatchPath(const std::string &path) { m_matchPath = path; }
+        void setItemAttributeName(const std::string &name) { m_itemAttribute = m_matchAttribute = name; }
```
Seems like you are relying on the order these methods may be called to change the state.  i.e. what if I did:

setMatchAttributeName("match");
setItemAttributeName("item");

?

Unless that's intentional, then in case setMatchAttributeName was already called I would do something like:
void setItemAttributeName(const std::string &name) 
{ 
m_itemAttribute = name; 
if (m_matchAttribute.empty())
    m_matchAttribute = name;
}

afishbeck commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -142,56 +171,148 @@ std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std
+
+    return pNewEnvNode;
+
+    /*pParentNode->addChild(pNewEnvNode);
```
this block of comment code is particularly ugly.

afishbeck commented on configuration/config2/EnvironmentMgr.cpp:
```diff
@@ -142,56 +171,148 @@ std::shared_ptr<EnvironmentNode> EnvironmentMgr::addNewEnvironmentNode(const std
+
+    //
+    // Send a create event now that it's been added to the environment
+    pCfgItem->findSchemaRoot()->processEvent("create", pNewEnvNode);
```
Wonder about the name of these events being "create".  Can you explain the meaning of create... when the comment above makes it sound like an "added".

afishbeck commented on configuration/config2/InsertableItem.cpp:
```diff
@@ -0,0 +1,39 @@
+        if (insertLimitType == "attribute")
+        {
+            std::string attributeName = pSchemaItem->getProperty("insertLimitData");
+            std::shared_ptr<SchemaValue> pSchemaValue = pSchemaItem->getAttribute(attributeName);
```
Again lots of variables that I think probably could be unique_ptr.

afishbeck commented on configuration/config2/InsertableItem.hpp:
```diff
@@ -0,0 +1,42 @@
+#include "SchemaItem.hpp"
+#include <vector>
+
+struct InsertItemLimitChoice
```
Do you use struct just to avoid a "public:" declaration?

kenrowland replied:
I suppose yes. It has been a style I have used that when a small simply all public class is necessary, I just use a struct to indicate that it's all public. I also use it when instantiated objects are treated more like data, such as items in a vector.

afishbeck commented on esp/services/ws_config2/ws_config2_binding.hpp:
```diff
@@ -27,4 +27,4 @@ class Cws_config2SoapBindingEx : public Cws_config2SoapBinding
 
 private:
-};
+};
```
missing carriage return.

kenrowland replied:
Fixed.

afishbeck reviewed (commented):
Generally looks ok.  Some comments.  There may be places where you should be using the "override" declaration when overriding virtual functions.  Didn't take the time to track down specific ones.

afishbeck commented:
@stuartort someone else should also review.  Especially details of c++ and std library usage.

rpastrana commented on configuration/config2/XMLEnvironmentLoader.cpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+
+void XMLEnvironmentLoader::parse(const pt::ptree &envTree, const std::shared_ptr<SchemaItem> &pConfigItem, std::shared_ptr<EnvironmentNode> &pEnvNode) const
```
didn't notice any member vars, is this method static by nature/should it be declared as such

kenrowland replied:
The intent is that you use this method when you have an instance of the object, not to be able to call it otherwise. In the future there could be use of class member variables.

rpastrana commented on esp/scm/ws_config2.ecm:
```diff
@@ -310,6 +329,20 @@ ESPresponse [exceptions_inline] GetTreeResponse
+ESPrequest FindNodesRequest
+{
+    string sessionId;
+    string startingNodeId("");   // optional stataring node for the find, "" for none
```
spelling

kenrowland replied:
Fixed.

rpastrana commented on esp/scm/ws_config2.ecm:
```diff
@@ -310,6 +329,20 @@ ESPresponse [exceptions_inline] GetTreeResponse
 };
 
 
+ESPrequest FindNodesRequest
```
It might be more common in hpcc code to call this FetchNodesRequest

kenrowland replied:
Unless there is a compelling reason, I will leave this as "find". A change to fetch will percolate all the way through the config manager.

rpastrana commented on esp/scm/ws_config2.ecm:
```diff
@@ -376,6 +409,10 @@ ESPservice [auth_feature("DEFERRED"),version("2.0"), default_client_version("2.0
+    ESPMethod
+        [
+            description("Find nodes matching the indicated path")
+        ] FindNodes(FindNodesRequest, FindNodesResponse);
```
FetchNodes?

kenrowland replied:
See previous comment. Leaving as find to keep same paradigm throughout all of config manager.

rpastrana commented on initfiles/componentfiles/config2xml/environment.xml:
```diff
@@ -0,0 +1,1111 @@
+<?xml version="1.0" encoding="utf-8"?>
```
what's this file used for? is it here for testing?

rpastrana commented on initfiles/componentfiles/config2xml/environment190.xml:
```diff
@@ -0,0 +1,6093 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Edited with ConfigMgr on ip  on 2018-03-23T18:29:29 -->
```
This file seems out of place in this folder, it prob shouldn't be checked in at all

rpastrana reviewed (changes requested):
@kenrowland several comments/questions. Noticed commented out code throughout. Remove 190env file

Commit:
HPCC-19216 Dynamically add <Properties> section from buildset

Add support to dynamically add <Properties> sections to
ESPService and ESPBinding sections in the environment

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b8806f1c632e5587310e2c5a6d31584ba961e17f
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  185 sec (00:03:05) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1161 sec (00:19:21) |   16 sec (00:00:16) | 1453 sec (00:24:13) |

afishbeck commented on configuration/config2/EnvironmentEvents.hpp:
```diff
@@ -0,0 +1,97 @@
+};
+
+
+class CreateEnvironmentEvent : public EnvironmentEvent
```
Or an EnvironmentCreateEvent?

afishbeck commented:
@kenrowland if you wait until the end of code review to rebase it will make things much easier for the reviewers.

kenrowland commented on configuration/config2/SchemaItem.cpp:
```diff
@@ -333,7 +367,7 @@ void SchemaItem::resetEnvironment()
 void SchemaItem::findSchemaValues(const std::string &path, std::vector<std::shared_ptr<SchemaValue>> &schemaValues)
 {
-    bool rootPath = path[0] == '/';
+    bool rootPath = path[0] == '/';   //todo: convert this to use the findSchemaRoot below
```
Comment no longer valid. Removed

kenrowland commented on configuration/config2/Utils.cpp:
```diff
@@ -0,0 +1,35 @@
+/*##############################################################################
```
I put this code in a common module for use by config2 manager. It's a static function that is used by a number of methods. Being located in the config2 directory I figured was enough to indicate it's for config2. If this should be moved to a system wide location, please tell me where. I am hesitant to move it unless it becomes part of a system lib because it would only be compiled and used by config2.

kenrowland reviewed (commented):
Comments addressed. New commit will follow shortly.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 47aee9fe3e8f61875168554a76e04830ca23922d
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  183 sec (00:03:03) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1165 sec (00:19:25) |   17 sec (00:00:17) | 1465 sec (00:24:25) |

kenrowland commented:
@afishbeck , @rpastrana 
Changes have been made. Please review latest commit.

Thanks.

rpastrana commented on configuration/config2/EnvironmentEventHandlers.cpp:
```diff
@@ -1,6 +1,6 @@
 /*##############################################################################
 
-HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems�.
+HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems�.
```
this cr symbol looks odd

kenrowland replied:
Fixed. I believe the issue is when I create a file in Windows, and push it. I think the character is different between Linux and Windows.

rpastrana reviewed (changes requested):
@kenrowland one minor issue

rpastrana commented:
Sending  back to @kenrowland

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: f2e2b754c6caada92f86fdb5e0b56458f1f37903
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |776 |776 |0 |
|test (thor) |707 |707 |0 |
|test (roxie) |820 |820 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   11 sec (00:00:11) |  188 sec (00:03:08) |   56 sec (00:00:56) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1161 sec (00:19:21) |   17 sec (00:00:17) | 1461 sec (00:24:21) |

richardkchapman commented:
@rpastrana Are you ok with this now? If so, will need squashing

kenrowland commented:
@rpastrana Renamed APIs to FetchXXX. Please review and approve.
@afishbeck Please approve if no further comments

afishbeck reviewed (approved):
@kenrowland looks ok

kenrowland commented:
@rpastrana If no further comments, please approve. Thanks

rpastrana commented:
@kenrowland latest commit looks reasonable, you can squash your commits.

kenrowland commented:
@richardkchapman Please merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 67b5147733e0e6a5629463c2e03306c7d2b003d2
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |784 |784 |0 |
|test (thor) |715 |715 |0 |
|test (roxie) |828 |828 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  188 sec (00:03:08) |   56 sec (00:00:56) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1171 sec (00:19:31) |   17 sec (00:00:17) | 1464 sec (00:24:24) |
